### PR TITLE
[example] Add SWE-agent example

### PR DESCRIPTION
**Pull request by yueming-yuan**
# Add SWE-Agent Training Example with Nemo-Gym Integration

## Overview
This PR adds an example for training SWE-Agent using Miles with NVIDIA's Nemo-Gym environment, SWE-Gym dataset, and SWE-bench evaluation.

## Changes Summary
- **New directory**: `examples/swe-agent/`
- **Submodules added**: 2 (Nemo-Gym and mini-swe-agent)

*Note: the implementation of this example is partially here, and partially in the above two submodules. Check the codes there for more details.

### 1. Core Integration Module (`generate_with_swe_agent.py`)
- **Custom generation function**: Integrates with Nemo-Gym's `/run` endpoint to execute agent trajectories
- **Token and mask builder**: Converts multi-turn agent conversations into training-ready tokens with proper loss masking
- **Reward function**: Extracts rewards from Gym environment responses
- **Dynamic filtering**: Filters out aborted samples to maintain training data quality
- **Agent metrics aggregation**: Collects and logs comprehensive agent performance metrics including:
  - Turn counts and tool call statistics
  - Time breakdowns (model query time, environment execution time, evaluation time)
  - Time ratios and overhead analysis
- **Custom rollout function**: Wraps base sglang rollout with agent-specific metric collection

### 2. Data Processing Utility (`download_and_process_data.py`)
- Downloads SWE-Gym dataset from HuggingFace
- Converts to Miles format with proper metadata structure
- Adds required `subset` and `split` fields for Gym API compatibility

### 3. Training Scripts
- **`run-qwen3-4b-instruct.sh`**: Production training script for Qwen3-4B-Instruct model

### 5. Submodules (`.gitmodules`)
- **nemo-gym**: The Gym environment provider, launching the environment/agent server. Fork with Miles-specific adaptations, added metrics computation, rollout status handling, sampling params handling, etc.
  - Branch: `miles-swe-agent`
  - URL: https://github.com/yueming-yuan/Gym
  
- **mini-swe-agent**: Lightweight SWE-agent implementation. Fork with Miles-specific adaptations, added metrics computation, rollout status handling, sampling params handling, etc.
  - Branch: `miles-swe-agent`
  - URL: https://github.com/yueming-yuan/nv-mini-swe-agent

### Integration Architecture
1. **Miles** (training framework) runs in one Docker container with GPU access
2. **Nemo-Gym** (environment server) runs in a separate Docker container
3. Communication via HTTP over a shared Docker network (`swe-net`)
4. Each SWE-Gym task spawns its own isolated Docker container for safe code execution

### Sample Status Handling
- `COMPLETED`: Agent successfully submitted a solution (`exit_status == "Submitted"`)
- `TRUNCATED`: Rollout exceeded limits (`RolloutTruncated`, `LimitsExceeded`, `CollapseContinued`)
- `ABORTED`: Other failures (filtered out from training via `dynamic_filter`)

### Loss Masking Strategy
- First 2 messages (system + user prompt): masked out (no loss)
- Assistant responses: loss applied (mask = 1)
- User/environment responses: masked out (mask = 0)

### CODE REVIEW

**Comment by zhaochenyang20** [comment]
Thanks. I will take a review on this. @yueming-yuan

**Comment by zijiexia** [comment]
I'm able to reproduce this example and you may find the training log [here](https://wandb.ai/zijie_xia-n-a/miles-swe-agent?nw=nwuserzijie_xia)

File: examples/swe-agent/generate_with_swe_agent.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        else:
+            response_text += content
+            if msg["role"] == "assistant":
+                loss_mask.extend([1] * len(msg_tokens))
+            else:
+                loss_mask.extend([0] * len(msg_tokens))
+
+    response_length = len(all_tokens) - prompt_length
+
+    return all_tokens, loss_mask, response_text, response_length
```
Comment (zhaochenyang20) [inline examples/swe-agent/generate_with_swe_agent.py:56]: The code performs frequent list extensions and manual mask handling within a Python loop. While currently on CPU, this becomes a GIL contention point.

File: examples/swe-agent/generate_with_swe_agent.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        if not content:
+            continue
+
+        msg_tokens = tokenizer(content, add_special_tokens=False)["input_ids"]
+        all_tokens.extend(msg_tokens)
+
+        if i < 2:
+            prompt_length += len(msg_tokens)
+        else:
+            response_text += content
```
Comment (zhaochenyang20) [inline examples/swe-agent/generate_with_swe_agent.py]: I replaced the string concatenation (`+=`) inside the loop with `list.append()` and `"".join()`.

strings are immutable, so repeated concatenation creates an O(N²) memory copy overhead.

File: examples/swe-agent/generate_with_swe_agent.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        content = msg.get("content", "")
+        if not content:
+            continue
+
+        msg_tokens = tokenizer(content, add_special_tokens=False)["input_ids"]
+        all_tokens.extend(msg_tokens)
+
+        if i < 2:
+            prompt_length += len(msg_tokens)
+        else:
```
Comment (zhaochenyang20) [inline examples/swe-agent/generate_with_swe_agent.py]: I decoupled the loop into prompt_msgs and response_msgs using slicing.

This removes the repeated if i < 2 branch check from the iteration loop, which is better for CPU branch prediction and makes the logic for "Prompt vs. Response" explicitly clear.
